bot/cogs/filtering.py

```diff
@@ -369,6 +398,48 @@ def mod_log(self) -> ModLog:
         except discord.errors.Forbidden:
             await channel.send(f"{filtered_member.mention} {reason}")
 
+    async def _scheduled_task(self, msg: dict) -> None:
+        """Delete an offensive message once its deletion date is reached."""
+        delete_at = dateutil.parser.isoparse(msg['delete_date'])
+
+        await wait_until(delete_at)
+        await self.delete_offensive_msg(msg)
+
+        self.cancel_task(msg['id'])
+
+    async def reschedule_offensive_msg_deletion(self) -> None:
+        """Get all the pending message deletion from the API and reschedule them."""
+        await self.bot.wait_until_ready()
+        response = await self.bot.api_client.get('bot/offensive-messages',)
+
+        now = datetime.datetime.utcnow()
+
+        for msg in response:
+            delete_at = dateutil.parser.isoparse(msg['delete_date']).replace(tzinfo=None)
+
+            if delete_at < now:
+                await self.delete_offensive_msg(msg)
+            else:
+                self.schedule_task(self.bot.loop, msg['id'], msg)
+
+    async def delete_offensive_msg(self, msg: Mapping[str, str]) -> None:
+        """Delete an offensive message, and then delete it from the db."""
+        try:
+            channel = self.bot.get_channel(msg['channel_id'])
+            if channel:
+                msg_obj = await channel.fetch_message(msg['id'])
+                await msg_obj.delete()
+        except NotFound:
+            log.info(
+                f"Tried to delete message {msg['id']}, but the message can't be found "
+                f"(it has been probably already deleted)."
+            )
+        except HTTPException as e:
+            log.warning(f"Failed to delete message {msg['id']}: status {e.status}")
```

Comment:
What kind of errors would cause this?

Currently, this error handling means that we would not remove the actual message, but do remove the entry from the database down below. So, the offensive message remains and will never be deleted. If these type of errors are mainly triggered by things like temporary disruptions at Discord's end, then it could be a good idea to keep the rows in the database and attempt removal at a later point in time.

However, that strategy would mean that completely "unremoveable" would linger in the database indefinitely.

I'm not sure. What do you think?

Reply:
Yeah, I put the except here in order to avoid uncaught exception, I think rescheduling them seems like the way to go. 
I'm not sure about unremoveable messages you're talking about though. Are you talking about errors not caused by a discord downtime, but by an actual error that will not solve with time? If so, yeah, it could be an issue. We could also delete them anyway if the deletion failed more than 3 times for instance.

Reply:
What if the message was manually deleted after the filter trigger? This is quite likely to happen a lot, and perhaps warning is not the right log level to use for this if we're going to be seeing it every time somone has deleted the message manually. I mean, as long as the message is gone, that's a good thing and no warning is necessary.

If we rescheduled these so that they would at least attempt to delete the message twice, then we could safely assume that if it fails twice, it's not due to some temporary disruption of the Discord service but because the message no longer exists. In which case, we could probably make this an INFO level log event.

This has the added benefit of not piling up unactionable sentry issues, since a warning would be sent to Sentry for us to deal with.

Reply:
> What if the message was manually deleted after the filter trigger? This is quite likely to happen a lot, 

There's a special `except NotFound` clause with an info-level log message. This clause seems to be specifically for messages that the bot either could not fetch for other reasons that `NotFound` (whatever that may be, it could be a temporary API fluke at Discord's end) or where the deletion action fails (for whatever reason).

Reply:
In that case, I suppose this is fine as a warning. I still think rescheduling is something we should probably do. At least once.

Reply:
@lemonsaurus there is an except block just above for `NotFound` with an appropriate info message

Reply:
I am facing a little issue. The `OffensiveMessage` model of the site database doesn’t allow PATCH requests, and the PR adding it was merged. I think we have two options:

---

### Option 1
Open a new PR to allow PATCH requests and a new deletion attempts field.
### Option 2
When the deletion needs to be retried, delete the current object from the database and create a new one. To check if we should still retry later, we could check if the message was posted `DELETION_TIME * 2` ago.

---

Which option do you prefer?

Reply:
A mix of the two sounds better to me. Make the API support PATCHing the `delete_date` (this is a very simple change to make with DRF). Make the client (i.e. the bot) compare the message's timestamp (keep in mind this can be calculated directly from the snowflake if needed) to determine if a deletion should be re-scheduled. A field for a deletion attempts just doesn't feel right to me.